format.go

```diff
@@ -34,130 +34,127 @@ func NewDefaultFormat(withTrace bool) Format {
 	var str string
-	if upErr.ErrChain != nil {
-		for _, eLink := range *upErr.ErrChain {
-			str += eLink.formatStr(format)
+	if upErr.ErrRoot.Msg != "" || len(upErr.ErrRoot.Stack) > 0 {
+		str += upErr.ErrRoot.formatStr(format)
+		if format.WithTrace && len(upErr.ErrChain) > 0 {
+			str += format.Sep
 		}
 	}
-	str += upErr.ErrRoot.formatStr(format)
+
+	for _, eLink := range upErr.ErrChain {
+		if !format.WithTrace {
+			str += format.Sep
+		}
+		str += eLink.formatStr(format)
+	}
+
 	if upErr.ExternalErr != "" {
 		str += fmt.Sprint(upErr.ExternalErr)
 	}
+
 	return str
 }
 
 // ToJSON returns a JSON formatted map for a given eris error.
 func (upErr *UnpackedError) ToJSON(format Format) map[string]interface{} {
-	if upErr == nil {
-		return nil
-	}
 	jsonMap := make(map[string]interface{})
-	if fmtRootErr := upErr.ErrRoot.formatJSON(format); fmtRootErr != nil {
-		jsonMap["error root"] = fmtRootErr
+	if upErr.ErrRoot.Msg != "" || len(upErr.ErrRoot.Stack) > 0 {
+		jsonMap["root"] = upErr.ErrRoot.formatJSON(format)
 	}
-	if upErr.ErrChain != nil {
+
+	if len(upErr.ErrChain) > 0 {
 		var wrapArr []map[string]interface{}
-		for _, eLink := range *upErr.ErrChain {
+		for _, eLink := range upErr.ErrChain {
 			wrapMap := eLink.formatJSON(format)
 			wrapArr = append(wrapArr, wrapMap)
 		}
-		jsonMap["error chain"] = wrapArr
+		jsonMap["wrap"] = wrapArr
 	}
+
 	if upErr.ExternalErr != "" {
-		jsonMap["external error"] = fmt.Sprint(upErr.ExternalErr)
+		jsonMap["external"] = fmt.Sprint(upErr.ExternalErr)
 	}
+
 	return jsonMap
 }
 
-func unpackRootErr(err *rootError) UnpackedError {
-	return UnpackedError{
-		ErrRoot: &ErrRoot{
-			Msg:   err.msg,
-			Stack: err.stack.get(),
-		},
-	}
+func (upErr *UnpackedError) unpackRootErr(err *rootError) {
+	upErr.ErrRoot.Msg = err.msg
+	upErr.ErrRoot.Stack = err.stack.get()
 }
 
-func unpackWrapErr(chain *[]ErrLink, err *wrapError) UnpackedError {
-	link := ErrLink{}
-	link.Frame = *err.frame.get()
-	link.Msg = err.msg
-	*chain = append(*chain, link)
-
-	e := UnpackedError{}
-	e.ErrChain = chain
+func (upErr *UnpackedError) unpackWrapErr(err *wrapError) {
+	// prepend links in stack trace order
+	link := ErrLink{Msg: err.msg}
+	wFrames := err.stack.get()
+	if len(wFrames) > 0 {
+		link.Frame = wFrames[0]
+	}
+	upErr.ErrChain = append([]ErrLink{link}, upErr.ErrChain...)
 
 	nextErr := err.Unwrap()
 	switch nextErr := nextErr.(type) {
-	case nil:
-		return e
 	case *rootError:
-		uErr := unpackRootErr(nextErr)
-		e.ErrRoot = uErr.ErrRoot
+		upErr.unpackRootErr(nextErr)
 	case *wrapError:
-		e = unpackWrapErr(chain, nextErr)
-	default:
-		e.ExternalErr = err.Error()
+		upErr.unpackWrapErr(nextErr)
 	}
-	return e
+
+	// insert the wrap frame into the root stack
+	upErr.ErrRoot.Stack.insertFrame(wFrames)
 }
 
 // ErrRoot represents an error stack and the accompanying message.
 type ErrRoot struct {
 	Msg   string
-	Stack []StackFrame
+	Stack Stack
 }
 
 func (err *ErrRoot) formatStr(format Format) string {
```

Comment:
move formatStr methods below the ToString

Reply:
rearrange the methods so that unexported util methods used in the exported ones are near them

Reply:
are they not already below them? or were you suggesting moving these directly below those methods? the ToString and ToJSON methods are part of a different type so i don't think that would make sense

Reply:
yeah I meant directly below them but I forgot we are following the same receiver methods stacked together convention.